llpc/context/llpcCompiler.cpp

```diff
@@ -728,6 +729,218 @@ template <typename T> static FormattedBytes formatBytesLittleEndian(ArrayRef<T>
+
+  if (result != Result::Success)
+    return result;
+
+  void *allocBuf = pipelineInfo->pfnOutputAlloc(pipelineInfo->pInstance, pipelineInfo->pUserData, candidateElf.size());
+  uint8_t *code = static_cast<uint8_t *>(allocBuf);
+  memcpy(code, candidateElf.data(), candidateElf.size());
+  pipelineOut->pipelineBin.codeSize = candidateElf.size();
+  pipelineOut->pipelineBin.pCode = code;
+  return result;
+}
+
+// =====================================================================================================================
+// Build the whole graphics pipeline. If missing elfPackage of a certain stage, we will build it first, and
+// link them to the full pipeline last.
+//
+// @param pipelineInfo : Info to build this shader module
+// @param [out] pipelineOut : Output of building this shader module
+// @param elfPackage : Early compiled elfPackage; it is an array which size is UnlinkedStageCount.
+// @returns : Result::Success if successful. Other return codes indicate failure.
+Result Compiler::buildGraphicsPipelineWithElf(const GraphicsPipelineBuildInfo *pipelineInfo,
+                                              GraphicsPipelineBuildOut *pipelineOut, const BinaryData *elfPackage) {
+  if (!pipelineInfo->pfnOutputAlloc)
+    return Result::ErrorInvalidPointer;
+
+  // clang-format off
+  SmallVector<const PipelineShaderInfo *, ShaderStageGfxCount> shaderInfo = {
+    &pipelineInfo->vs,
+    &pipelineInfo->tcs,
+    &pipelineInfo->tes,
+    &pipelineInfo->gs,
+    &pipelineInfo->fs,
+  };
+  // clang-format on
+
+  ElfPackage elf[UnlinkedStageCount] = {};
+
+  MetroHash::Hash cacheHash = {};
+  MetroHash::Hash pipelineHash = {};
+  cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+  pipelineHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true);
+
+  GraphicsContext graphicsContext(m_gfxIp, pipelineInfo, &pipelineHash, &cacheHash);
+  Context *context = acquireContext();
+  context->attachPipelineContext(&graphicsContext);
+
+  // Release context.
+  auto onExit = make_scope_exit([&] { releaseContext(context); });
+
+  Result result = Result::Success;
+  for (unsigned idx = 0; idx < UnlinkedStageCount; idx++) {
+    if (elfPackage[idx].pCode) {
+      auto data = reinterpret_cast<const char *>(elfPackage[idx].pCode);
+      elf[idx].assign(data, data + elfPackage[idx].codeSize);
+    } else {
+      graphicsContext.setUnlinked(true);
+      result = buildUnlinkedShaderInternal(context, shaderInfo, static_cast<UnlinkedShaderStage>(idx), elf[idx],
+                                           pipelineOut->stageCacheAccesses);
+      if (result != Result::Success)
+        return result;
+    }
+  }
+
+  // Link
+  graphicsContext.setUnlinked(false);
+  ElfPackage pipelineElf;
+  bool hasError = false;
+  context->setDiagnosticHandler(std::make_unique<LlpcDiagnosticHandler>(&hasError));
+  hasError |= !linkRelocatableShaderElf(elf, &pipelineElf, context);
```

Comment:
I there a plan for how the handle the glue code?  This currently compiles the glue code at link time.  Will we be exposing an interface to compile the vertex fetch shader and color export shader as well?